src/Campaigns/Migrations/MigrateFormsToCampaignForms.php

```diff
@@ -29,66 +26,153 @@ public static function id(): string
      */
     public static function timestamp(): int
     {
-        return strtotime('2024-08-21');
+        return strtotime('2024-08-26 00:00:01');
     }
 
     /**
      * @unreleased
      * @inheritDoc
+     * @throws \Exception
      */
     public function run()
     {
-        foreach(DonationForm::query()->getAll() ?? [] as $form) {
-            $this->createParentCampaignForDonationForm($form);
-        }
+        DB::transaction(function() {
+            try {
+                array_map([$this, 'createCampaignForForm'], $this->getFormData());
+                array_map([$this, 'addUpgradedFormToCampaign'], $this->getUpgradedFormData());
```

Comment:
As a note, inserting multiple rows at once has to be batched. I think the limit is somewhere around 10k at once. We are using a transaction here, so that will help quite a bit.

This may be entirely fine as it is, to be clear, I'm just sharing thoughts. My typical preference for these sorts of migrations is to use INSERT INTO with a subquery — and let the DB engine handle the whole thing.

Reply:
Good point, @JasonTheAdams. This could certainly be improved by inserting multiple rows in a single query, particularly for the upgraded forms.

Our `DB::insert` method maps to `wpdb::insert`, which only supports inserting a single row. I'm not seeing any `wpdb` functions for inserting multiple rows.

I'll look into using a sub-query.

Reply:
@JasonTheAdams I was looking into manually adding the INSERT INTO with a subquery for the migration as a step towards updating the query builder, but I think it would be better to just go ahead and update the query builder and refactor this work accordingly in a subsequent PR.

Reply:
I'm totally good with that. 👍

Reply:
@JasonTheAdams for follow up see https://github.com/impress-org/givewp/pull/7538.